pytensor/tensor/nlinalg.py

```diff
@@ -582,26 +585,87 @@ def infer_shape(self, fgraph, node, shapes):
         else:
             return [s_shape]
 
+    def L_op(
+        self,
+        inputs: Sequence[Variable],
+        outputs: Sequence[Variable],
+        output_grads: Sequence[Variable],
+    ) -> list[Variable]:
+        (A,) = inputs
+        A = cast(ptb.TensorVariable, A)
 
-def svd(a, full_matrices: bool = True, compute_uv: bool = True):
+        if not self.compute_uv:
+            (S_grad,) = output_grads
+            S_grad = cast(ptb.TensorVariable, S_grad)
+
+            # Need U and V so do the whole svd anyway...
+            [u, s, v] = svd(A, full_matrices=False, compute_uv=True)  # type: ignore
```

Comment:
Then we should have a rewrite that drops SVD without `compute_uv=True` when both exist in the graph

Reply:
What is "both" in this sentence?

Reply:
If there's a node with True and one with False but the same inputs we would replace the False with the True (if any of the other outputs actually have clients)

Reply:
Is it possible to check if the `Op` is on a gradient graph? We should just always rewrite it to `compute_uv = True` if so -- that would be best I think.

Reply:
There's no way to know if a graph is a gradient graph or not, but why do you need that? This sounds like a simple merge optimization. If there are two SVDs with the same inputs and different compute_uv, use the one that computes it. Conversely if there is only one SVD with compute uv, but those aren't used, rewrite it to compute_uv = False?

Reply:
Is there an example of a rewrite like this I can follow?

Reply:
I think there's a MaxAndArgmax that specializes when only one of the two outputs is used (argmax is used in the grad of max).

I don't think there's one in the other direction because by default we always give the user the MaxAndArgmax so it's only necessary to go in one direction.

Reply:
Do you still think I should do a rewrite for the `compute_uv = False` case? It would mean we just never trigger the first gradient case, and always end up in the `is_disconnected =[True, False, True]` path of the `compute_uv = True` case.

Reply:
Grad is eager so it usually has to handle the cases before rewrites take place.

The rewrite I had in mind doesn't affect which form it is because it's not grad specific (although it always applies to the grad generated by the first branch)

The idea is if you have two nodes in the same graph with SVD(x, compute_uv=False) and SVD(x, compute_uv=True) you can replace the output of the first node with the second output of the second node.

Besides that you could have a rewrite when the matrices are not used to replace SVD(x, compute_uv=True) by SVD(x, compute_uv=False)

None of these should affect how you implemented the grad, but they can make the final graph more efficient?

Reply:
All great ideas. But do you want them as blockers for this PR, or can I file them under #573 and hope the GSOC money comes through? :D

Reply:
Definitely not blockers and nice beginner friendly issues

Reply:
They are also useful as a template, as there was some discussion on the past about making grad Op implementations more complex (like jax has to do) to avoid redundant computations between forward and grad passes.

This shows how PyTensor can handle it easily enough with rewrites under the current machinery

Reply:
OK in that case I'd rather leave it for another day. I'll open a new issue about it after this PR is merged

Reply:
> OK in that case I'd rather leave it for another day. I'll open a new issue about it after this PR is merged 

Can open already